src/dev/encoder.cpp

```diff
@@ -0,0 +1,127 @@
+#include <EVT/dev/encoder.hpp>
+
+namespace EVT::core::DEV {
+
+Encoder::Encoder(IO::GPIO& a, IO::GPIO& b, uint64_t range, uint64_t initialPosition, bool rollOver)
+    : a(a), b(b), range(range), position(initialPosition), rollOver(rollOver) {
+    if (position > range) {
+        position = range;
+    }
+    //setting instance variables
+    currentRelPos = readPinValues();
+    interruptChange = 0;
+    recentInterruptTime = 0;
+}
+
+void Encoder::aInterruptHandler(IO::GPIO* pin) {
+    if (time::millis() - recentInterruptTime < INTERRUPTCOOLDOWN) {
+        return;
+    }
+    recentInterruptTime = time::millis();
+    int8_t change;
+    switch (currentRelPos) {
+    case 0:
+        change = 1;
+        currentRelPos = 1;
+        break;
+    case 1:
+        change = -1;
+        currentRelPos = 0;
+        break;
+    case 2:
+        change = 1;
+        currentRelPos = 3;
+        break;
+    case 3:
+        change = -1;
+        currentRelPos = 2;
+        break;
+    }
+    log::LOGGER.log(log::Logger::LogLevel::DEBUG, "aInterrupt Called, position: %d", currentRelPos);
+    interruptChange += change;
+}
+
+void Encoder::bInterruptHandler(IO::GPIO* pin) {
+    if (time::millis() - recentInterruptTime < INTERRUPTCOOLDOWN) {
+        return;
+    }
+    recentInterruptTime = time::millis();
+    int8_t change;
+    switch (currentRelPos) {
+    case 0:
+        change = -1;
+        currentRelPos = 3;
+        break;
+    case 1:
+        change = 1;
+        currentRelPos = 2;
+        break;
+    case 2:
+        change = -1;
+        currentRelPos = 1;
+        break;
+    case 3:
+        change = 1;
+        currentRelPos = 0;
+        break;
+    }
+    log::LOGGER.log(log::Logger::LogLevel::DEBUG, "bInterrupt Called, position: %d", currentRelPos);
+    interruptChange += change;
+}
+
+uint64_t Encoder::getPosition() const {
+    return position;
+}
+
+void Encoder::update() {
+    int64_t change = interruptChange;
+    interruptChange = 0;
+    changePosition(change);
+}
+
+int8_t Encoder::readPinValues() {
+    bool aPos = (bool) a.readPin();
+    bool bPos = (bool) b.readPin();
+    int8_t newPos = convertPinValuesToPosition(aPos, bPos);
+    return newPos;
+}
+
+int8_t Encoder::convertPinValuesToPosition(bool a, bool b) {
```

Comment:
Is this function used anywhere else then on line 85? If not could it just be moved into that function instead of an entire standalone function?

Reply:
Yeah, this logic is only really going to be useful in `readPinValues`, so I think it makes more sense to be in there

Reply:
Moved it!